ktor-server/ktor-server-netty/jvm/src/io/ktor/server/netty/cio/NettyResponsePipeline.kt

```diff
@@ -6,138 +6,94 @@ package io.ktor.server.netty.cio
+     */
+    private val isWriteOccurred: AtomicBoolean = AtomicBoolean(false)
+
+    /**
+     * Represents promise which is marked as success when the last read request is handled.
+     * Marked as fail when last read request is failed.
+     * Default value is success on purpose to start first request handle
+     */
+    private var previousCallHandled: ChannelPromise = context.newPromise().also {
+        it.setSuccess()
     }
 
-    private var encapsulation: WriterEncapsulation = initialEncapsulation
-
-    fun ensureRunning() {
-        responses.start()
-    }
-
-    private suspend fun processJobs() {
-        while (true) {
-            fill()
-            if (running.isEmpty()) break
-            processElement(running.removeFirst())
-        }
-
-        if (encapsulation.requiresContextClose) {
-            dst.close()
-        }
-    }
-
-    private suspend fun fill() {
-        tryFill()
-        if (running.isEmpty()) {
-            return fillSuspend()
+    /**
+     * Flush if there is some unflushed data, nothing to read from channel and no active requests
+     */
+    internal fun flushIfNeeded() {
+        if (isWriteOccurred.get() && isChannelReadComplete.get() && activeRequests.get() == 0L) {
+            context.flush()
+            isWriteOccurred.set(false)
         }
     }
 
-    private fun tryFill() {
-        while (isNotFull()) {
-            if (!pollReady()) {
-                tryStart()
-                dst.read()
-                break
-            }
+    internal fun processResponse(call: NettyApplicationCall) {
+        call.previousCallFinished = previousCallHandled
+        call.callFinished = context.newPromise()
+        previousCallHandled = call.callFinished
 
-            tryStart()
-        }
+        processElement(call)
     }
 
-    private suspend fun fillSuspend() {
-        if (running.isEmpty()) {
-            @OptIn(ExperimentalCoroutinesApi::class)
-            val e = incoming.receiveCatching().getOrNull()
+    private fun processElement(call: NettyApplicationCall) {
+        val previousCallFinished = call.previousCallFinished
 
-            if (e != null && e.ensureRunning()) {
-                running.addLast(e)
-                tryFill()
-            }
-        }
-    }
-
-    private fun pollReady(): Boolean {
-        for (index in 1..(readyQueueSize - ready.size)) {
-            val e = incoming.tryReceive().getOrNull() ?: return false
-            ready.addLast(e)
+        if (previousCallFinished.isDone && !previousCallFinished.isSuccess) {
+            processCallFailed(call, previousCallFinished.cause())
+            return
         }
-        return true
-    }
 
-    private fun tryStart() {
-        while (ready.isNotEmpty() && running.size < runningQueueSize) {
-            val e = ready.removeFirst()
-            if (e.ensureRunning()) {
-                running.addLast(e)
-            } else {
-                break
+        call.response.responseFlag.addListener {
```

Comment:
Please, consider renaming `responseFlag` (already discussed).
Other than that, maybe moving this whole part to a separate function with an explanative name and optionally a comment will be also helpful.